src/Microsoft.OData.Core/UriParser/SemanticAst/EachSegment.cs

```diff
@@ -0,0 +1,90 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="EachSegment.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.UriParser
+{
+    #region Namespaces
+    using Microsoft.OData.Edm;
+    #endregion Namespaces
+
+    /// <summary>
+    /// A segment representing $each in a path.
+    /// </summary>
+    public sealed class EachSegment : ODataPathSegment
+    {
+        /// <summary>
+        /// The <see cref="IEdmType"/> of the resource that this <see cref="EachSegment"/> represents.
+        /// </summary>
+        private readonly IEdmType edmType;
+
+        /// <summary>
+        /// Build a segment representing each.
+        /// </summary>
+        /// <param name="navigationSource">The entity collection that this set-based operation applies to.</param>
+        /// <param name="targetEdmType">Target type for the entity being referenced.</param>
+        /// <exception cref="System.ArgumentNullException">Throws if any input parameter is null.</exception>
+        /// <remarks>$filter cannot be applied on singletons.</remarks>
+        public EachSegment(IEdmNavigationSource navigationSource, IEdmType targetEdmType)

```

Comment:
>IEdmType targetEdmType [](start = 66, length = 22)

when would targetEdmType be different from navigationSource.Type?

Reply:
I don't have the in-depth reason off the top of my head but from previous tests, I found that they're different when the previous segment is either a TypeSegment or FilterSegment.

```
EachSegment eachSegment = new EachSegment(
                lastNavigationSource,
                prevSegment is TypeSegment || prevSegment is FilterSegment ? prevSegment.TargetEdmType : null);
```

This is in the ODataPathParser.cs. Please let me know if we should dig deeper on whether we should change this design.

Reply:
Got it.

Rather than make targetEdmType parameter nullable, wouldn't it be simpler to always pass in the targetEdmType and simply set this.TargetEdmType = targetEdmType in the constructor (rather than check for null)?  

In the code above, couldn't you just pass in prevSegment.TargetEdmType for all cases (presumably, if the previous segment was the navigation source, then lastNavigationSource.EntityType() == prevSegment.TargetEdmType?)

---
In reply to: [224878387](https://github.com/OData/odata.net/pull/1110#discussion_r224878387) [](ancestors = 224878387)

Reply:
Great point, I'll move the null check out of the constructor and make it not nullable. Additionally the check is made prior to construction. There may be some other path segments that are following this pattern, where the check is done inside the constructor; I think your suggestion is cleaner.